test/e2e/generator/generate.go

```diff
@@ -279,3 +316,99 @@ func generateLightNode(r *rand.Rand, startAt int64, providers []string) *e2e.Man
 func ptrUint64(i uint64) *uint64 {
 	return &i
 }
+
+// Parses strings like "v0.34.21:1,v0.34.22:2" to represent two versions
+// ("v0.34.21" and "v0.34.22") with weights of 1 and 2 respectively.
+func parseWeightedVersions(s string) (weightedChoice, error) {
+	wc := make(weightedChoice)
+	wvs := strings.Split(strings.TrimSpace(s), ",")
+	for _, wv := range wvs {
+		parts := strings.Split(strings.TrimSpace(wv), ":")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("unexpected weight:version combination: %s", wv)
+		}
+		ver := strings.TrimSpace(parts[0])
+		wt, err := strconv.Atoi(strings.TrimSpace(parts[1]))
+		if err != nil {
+			return nil, fmt.Errorf("unexpected weight \"%s\": %v", parts[1], err)
+		}
+		if wt < 1 {
+			return nil, errors.New("version weights must be >= 1")
+		}
+		wc[ver] = uint(wt)
+	}
+	return wc, nil
+}
+
+// Extracts the latest release version from the given Git repository. Uses the
+// current version of Tendermint Core to establish the "major" version
+// currently in use.
+func gitRepoLatestReleaseVersion(gitRepoDir string) (string, error) {
+	opts := &git.PlainOpenOptions{
+		DetectDotGit: true,
+	}
+	r, err := git.PlainOpenWithOptions(gitRepoDir, opts)
+	if err != nil {
+		return "", err
+	}
+	tags := make([]string, 0)
+	tagObjs, err := r.TagObjects()
+	if err != nil {
+		return "", err
+	}
+	err = tagObjs.ForEach(func(tagObj *object.Tag) error {
+		tags = append(tags, tagObj.Name)
+		return nil
+	})
+	if err != nil {
+		return "", err
+	}
+	return findLatestReleaseTag(version.TMCoreSemVer, tags)
+}
+
+func findLatestReleaseTag(baseVer string, tags []string) (string, error) {
+	baseSemVer, err := semver.NewVersion(strings.Split(baseVer, "-")[0])
+	if err != nil {
+		return "", fmt.Errorf("failed to parse base version \"%s\": %v", baseVer, err)
+	}
+	compVer := fmt.Sprintf("%d.%d", baseSemVer.Major(), baseSemVer.Minor())
+	// Build our version comparison string
+	// See https://github.com/Masterminds/semver#caret-range-comparisons-major for details
+	compStr := "^ " + compVer
+	verCon, err := semver.NewConstraint(compStr)
+	if err != nil {
+		return "", err
+	}
+	var latestVer *semver.Version
+	for _, tag := range tags {
+		if !strings.HasPrefix(tag, "v") {
+			continue
+		}
+		curVer, err := semver.NewVersion(tag)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse tag \"%s\" as semantic version: %v", tag, err)
```

Comment:
You might want to continue here... if someone tags some commit with an arbitrary tag starting with "v", it would break this logic

Reply:
Makes sense! Will change this.